core/control-panel/impl/src/controllers/http.rs

```diff
@@ -1,115 +1,97 @@
 use crate::core::metrics::METRIC_ACTIVE_USERS;
-use crate::services::{UserService, USER_SERVICE};
+use crate::core::middlewares::use_canister_call_metric;
+use crate::services::USER_SERVICE;
 use crate::{
-    core::ic_cdk::api::{canister_balance, time},
+    core::ic_cdk::api::{
+        canister_balance, data_certificate, print, set_certified_data, time, trap,
+    },
     SERVICE_NAME,
 };
-use ic_cdk_macros::query;
-use lazy_static::lazy_static;
-use orbit_essentials::api::{HeaderField, HttpRequest, HttpResponse};
-use orbit_essentials::http::add_skip_certification_headers;
+use ic_cdk_macros::{query, update};
+use ic_http_certification::{HttpRequest, HttpResponse};
+use orbit_essentials::api::ApiResult;
+use orbit_essentials::http::{certify_assets, serve_asset};
 use orbit_essentials::metrics::with_metrics_registry;
-use std::sync::Arc;
+use orbit_essentials::with_middleware;
 
-// Canister entrypoints for the controller.
-#[query(name = "http_request", decoding_quota = 10000)]
-async fn http_request(request: HttpRequest) -> HttpResponse {
-    let mut resp = CONTROLLER.router(request).await;
-    add_skip_certification_headers(&mut resp);
-    resp
+// no-op endpoint to refresh cycles balance in metrics
+#[update]
+async fn ping() {
+    let _ = do_ping().await;
 }
 
-// Controller initialization and implementation.
-lazy_static! {
-    static ref CONTROLLER: HttpController = HttpController::new(Arc::clone(&USER_SERVICE));
+// it is important to collect metrics here to refresh cycles balance in metrics
+#[with_middleware(tail = use_canister_call_metric("ping", &result))]
+async fn do_ping() -> ApiResult<()> {
+    Ok(())
 }
 
-#[derive(Debug)]
-pub struct HttpController {
-    user_service: Arc<UserService>,
-}
-
-impl HttpController {
-    fn new(user_service: Arc<UserService>) -> Self {
-        Self { user_service }
+#[query(decoding_quota = 10000)]
+fn http_request(req: HttpRequest) -> HttpResponse {
+    let res = serve_asset(&req, data_certificate());
+    match res {
+        Ok(response) => response,
+        Err(err) => trap(err),
     }
+}
 
-    async fn router(&self, request: HttpRequest) -> HttpResponse {
-        if request.url == "/metrics" || request.url == "/metrics/" {
-            return self.metrics(request).await;
+// Certification
+pub fn certify_metrics() {
+    // Trigger active users metric update.
+    METRIC_ACTIVE_USERS.with(|metric| metric.borrow_mut().refresh(time()));
+
+    // Add dynamic metrics, dropped after the request since query calls don't save state changes.
+    with_metrics_registry(SERVICE_NAME, |registry| {
+        registry
+            .gauge_mut(
+                "canister_cycles_balance",
+                "cycles balance available to the canister",
+            )
+            .set(canister_balance() as f64);
+    });
+    with_metrics_registry(SERVICE_NAME, |registry| {
+        registry
+            .gauge_mut(
+                "metrics_timestamp",
```

Comment:
Alternatively the metrics registry could be adapted to use the prometheus format which optionally can include timestamp per each metric (check [here](https://github.com/prometheus/docs/blob/main/content/docs/instrumenting/exposition_formats.md#text-format-details))

Reply:
I don't see how this can be done with the Rust prometheus crate.